File: src/logic/rent.ts
import { SpaceType, type Space, type Player } from '../types/game';

export function calculatePropertyRent(space: Space, dice?: [number, number]): number {
  if (!space.rent) return 0;

  if (space.type === SpaceType.Utility) {
    return calculateUtilityRent(space, dice);
  }

  if (space.type === SpaceType.Railroad) {
    return calculateRailroadRent(space);
  }

  const houseIndex = space.houses === 5 ? (space.rent?.length ?? 0) - 1 : space.houses;
  if (houseIndex >= (space.rent?.length ?? 0)) return space.rent?.[space.rent.length - 1] ?? 0;
  return space.rent?.[houseIndex] ?? 0;
}

function calculateRailroadRent(space: Space): number {
  const count = getRailroadCount(space);
  return space.rent?.[count - 1] ?? 25;
}

function calculateUtilityRent(space: Space, dice?: [number, number]): number {
  const total = (dice?.[0] ?? 0) + (dice?.[1] ?? 0);
  const count = getUtilityCount(space);
  return count === 2 ? total * 10 : total * 4;
}

export function getRailroadCount(space: Space): number {
  return (space as unknown as { _railroadCount: number })._railroadCount ?? 1;
}

export function setRailroadCount(space: Space, count: number): Space {
  return { ...space, _railroadCount: count } as Space;
}

function getUtilityCount(space: Space): number {
  return (space as unknown as { _utilityCount: number })._utilityCount ?? 1;
}

export function calculateRailroadRentFromBoard(ownerId: number, board: Space[], spaceId: number): number {
  const railroads = board.filter(
    (s) => s.type === SpaceType.Railroad && s.owner === ownerId
  );
  const count = railroads.length;
  const space = board[spaceId];
  return space.rent?.[count - 1] ?? 25;
}

export function calculateUtilityRentFromBoard(ownerId: number, board: Space[], _spaceId: number, dice: [number, number]): number {
  const utilities = board.filter(
    (s) => s.type === SpaceType.Utility && s.owner === ownerId
  );
  const count = utilities.length;
  const total = dice[0] + dice[1];
  return count === 2 ? total * 10 : total * 4;
}

export function getPlayerTotalAssets(player: Player, board: Space[]): number {
  let total = player.money;
  for (const pid of player.properties) {
    const space = board[pid];
    if (!space) continue;
    if (space.mortgaged) continue;
    total += (space.price ?? 0) / 2;
    if (space.type === SpaceType.Property || space.type === SpaceType.Railroad || space.type === SpaceType.Utility) {
      total += (space.houseCost ?? 0) * space.houses;
    }
  }
  return total;
}

File: backend/internal/usecase/gameusecase/rent.go
package gameusecase

import (
	"monopoly-game-backend/internal/data"
	"monopoly-game-backend/internal/entity"
)

// CalculatePropertyRent returns the rent for a space based on its type and house count.
// For properties, it uses the rent array indexed by houses.
// For railroads and utilities, it dispatches to specialized calculators.
func CalculatePropertyRent(space entity.Space, dice *[2]int) int {
	if space.Rent == nil {
		return 0
	}

	switch space.Type {
	case entity.SpaceTypeUtility:
		return calculateUtilityRent(space, dice)
	case entity.SpaceTypeRailroad:
		return calculateRailroadRent(space)
	}

	houseIndex := space.Houses
	if houseIndex == data.MaxHouses {
		houseIndex = len(space.Rent) - 1
	} else if houseIndex >= len(space.Rent) {
		houseIndex = len(space.Rent) - 1
	}
	if houseIndex < 0 {
		houseIndex = 0
	}
	return space.Rent[houseIndex]
}

func calculateRailroadRent(space entity.Space) int {
	count := GetRailroadCount(space)
	idx := count - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(space.Rent) {
		return 25
	}
	return space.Rent[idx]
}

func calculateUtilityRent(space entity.Space, dice *[2]int) int {
	total := 0
	if dice != nil {
		total = dice[0] + dice[1]
	}
	count := GetUtilityCount(space)
	if count == 2 {
		return total * 10
	}
	return total * 4
}

// GetRailroadCount returns the railroad count for a space, defaulting to 1.
func GetRailroadCount(space entity.Space) int {
	if space.RailroadCount <= 0 {
		return 1
	}
	return space.RailroadCount
}

// SetRailroadCount returns a copy of the space with the given railroad count.
func SetRailroadCount(space entity.Space, count int) entity.Space {
	space.RailroadCount = count
	return space
}

// GetUtilityCount returns the utility count for a space, defaulting to 1.
func GetUtilityCount(space entity.Space) int {
	if space.UtilityCount <= 0 {
		return 1
	}
	return space.UtilityCount
}

// SetUtilityCount returns a copy of the space with the given utility count.
func SetUtilityCount(space entity.Space, count int) entity.Space {
	space.UtilityCount = count
	return space
}

// CalculateRailroadRentFromBoard calculates railroad rent by scanning the board
// for railroads owned by the given player.
func CalculateRailroadRentFromBoard(ownerID int, board []entity.Space, spaceID int) int {
	if spaceID < 0 || spaceID >= len(board) {
		return 25
	}
	count := 0
	for _, s := range board {
		if s.Type == entity.SpaceTypeRailroad && s.Owner != nil && *s.Owner == ownerID {
			count++
		}
	}
	idx := count - 1
	if idx < 0 {
		idx = 0
	}
	space := board[spaceID]
	if space.Rent == nil {
		return 25
	}
	if idx >= len(space.Rent) {
		return 25
	}
	return space.Rent[idx]
}

// CalculateUtilityRentFromBoard calculates utility rent by scanning the board
// for utilities owned by the given player.
func CalculateUtilityRentFromBoard(ownerID int, board []entity.Space, _spaceID int, dice [2]int) int {
	count := 0
	for _, s := range board {
		if s.Type == entity.SpaceTypeUtility && s.Owner != nil && *s.Owner == ownerID {
			count++
		}
	}
	total := dice[0] + dice[1]
	if count == 2 {
		return total * 10
	}
	return total * 4
}

// IsMonopoly checks if the given player owns all properties of the same color as the space.
func IsMonopoly(ownerID int, board []entity.Space, space entity.Space) bool {
	if space.Type != entity.SpaceTypeProperty || space.Color == nil {
		return false
	}
	color := *space.Color
	groupSize := 0
	for _, s := range board {
		if s.Type == entity.SpaceTypeProperty && s.Color != nil && *s.Color == color {
			groupSize++
			if s.Owner == nil || *s.Owner != ownerID {
				return false
			}
		}
	}
	return groupSize > 0
}

// CalculatePlayerTotalAssets returns the total assets (money + half property prices + house investments).
func CalculatePlayerTotalAssets(player entity.Player, board []entity.Space) int {
	total := player.Money
	for _, pid := range player.Properties {
		if pid < 0 || pid >= len(board) {
			continue
		}
		space := board[pid]
		if space.Mortgaged {
			continue
		}
		if space.Price != nil {
			total += *space.Price / 2
		}
		switch space.Type {
		case entity.SpaceTypeProperty, entity.SpaceTypeRailroad, entity.SpaceTypeUtility:
			total += data.GetTotalHouseInvestment(space)
		}
	}
	return total
}

// CalculatePlayerNetWorth returns the net worth (money + full property prices + house investments).
func CalculatePlayerNetWorth(player entity.Player, board []entity.Space) int {
	total := player.Money
	for _, pid := range player.Properties {
		if pid < 0 || pid >= len(board) {
			continue
		}
		space := board[pid]
		if space.Price != nil {
			total += *space.Price
		}
		total += data.GetTotalHouseInvestment(space)
	}
	return total
}
